README.md

```diff
@@ -174,13 +192,13 @@ class MyHandler
 end
 ```
 
-Finally, it is also possible to preprocess the message payload before consuming it using the `before_consume` hook which is invoked before `.around_consume` and `#consume`. The result of this operation will be assigned to payload, so it is important to return the modified payload. This can be very useful, for example if you want a single point of decoding Avro messages and want the payload as a hash instead of a binary.
+Finally, it is also possible to preprocess the message payload before consuming it using the `before_consume` hook which is invoked before `#around_consume` and `#consume`. The result of this operation will be assigned to payload, so it is important to return the modified payload. This can be very useful, for example if you want a single point of decoding Avro messages and want the payload as a hash instead of a binary.
 
 ```ruby
 class MyHandler
   include Phobos::Handler
 
-  def before_consume(payload)
+  def before_consume(payload, metadata)
```

Comment:
This is a breaking change. I would like to have access to the metadata inside of before_consume for logging purposes (right now i'm digging into the Phobos.config to get that information). My preference would be to instantiate the handler with the metadata so there's no need to pass it as an arg to `before_consume`, `around_consume`, or `consume`:

```
def process_message(payload)
  instrument('listener.process_message', @metadata) do
    handler = @listener.handler_class.new(@metadata)
    consume_block = Proc.new { handler.consume(preprocessed_payload) }

    if @listener.handler_class.respond_to?(:around_consume)
      # around_consume class method implementation
      @listener.handler_class.around_consume(preprocessed_payload, &consume_block)
    else
      # around_consume instance method implementation
      handler.around_consume(preprocessed_payload, &consume_block)
    end
  end
end
```
However, this is a larger breaking change. Open other perspectives on this.

Reply:
There are 2 topics of discussion here.

### This code/The breaking change 

Well, I guess it will be a bumpy ride since we might want to remove `before_consume` in the future, but oh well, it was our bad for not adding metadata to it, in the first place.

### About the suggestion/preference
At first glance it looks clean, I like it. 

But it could be asked by someone else, why not also send payload as argument to initialize? But we happen to know that sometimes we want to change payload (decode and whatnot). So... could a third or fourth party potentially want to change/mutate metadata? 🤔 Also, in practice, will it not look a bit weird that we have methods with `metadata` appearing from nowhere, such as:

```ruby
def around_consume(payload)
  payload = decode_payload(payload)
  logger.log(payload: payload)
  logger.log(metadata: metadata)
end
```

Where did metadata come from? Looks a bit magic. Finally, since we discussed yielding payload to from around_consume, maybe we want to yield metadata as well here to not make the same mistake as we did (ignoring metadata)

Maybe you can create a new Issue for passing the `@metadata` to the initializer of the handler class and let's continue the discussion there.

Reply:
Sounds good! Yes i realize this PR really opens up a whole can of worms on discrete but related design decisions. I will work on a new issue for discussing the instantiation of the handler.